src/Microsoft.AspNetCore.Server.Testing/Deployers/NginxDeployer.cs

```diff
@@ -0,0 +1,139 @@
+
+            var exitToken = StartSelfHost(new Uri(redirectUri));
+
+            SetupNginx(redirectUri, uri);
+
+            // Wait for App to be loaded since Nginx returns 502 instead of 503 when App isn't loaded
+            // Target actual address to avoid going through Nginx proxy
+            var httpClient = new HttpClient();
+            var response = RetryHelper.RetryRequest(() =>
+            {
+                return httpClient.GetAsync(redirectUri);
+            }, Logger).Result;
+
+            if (!response.IsSuccessStatusCode)
+            {
+                throw new InvalidOperationException("Deploy failed");
+            }
+            
+            // Start timer
+            StartTimer();
+
+            if (DeploymentParameters.PublishApplicationBeforeDeployment)
+            {
+                DotnetPublish();
+            }
+            
+            return new DeploymentResult
+            {
+                WebRootLocation = DeploymentParameters.ApplicationPath,
+                DeploymentParameters = DeploymentParameters,
+                ApplicationBaseUri = uri.ToString(),
+                HostShutdownToken = exitToken
+            };
+        }
+
+        private void SetupNginx(string redirectUri, Uri originalUri)
+        {
+            // create nginx.conf and sites-enabled/default for redirect
+            File.WriteAllText($"{DeploymentParameters.ApplicationPath}/nginx.conf",
+$@"user {Environment.GetEnvironmentVariable("LOGNAME")};
+worker_processes 4;
+pid /run/nginx.pid;
+
+events {{
+    worker_connections 768;
+}}
+
+http {{
+    sendfile on;
+    tcp_nopush on;
+    tcp_nodelay on;
+    keepalive_timeout 10;
+    types_hash_max_size 2048;
+
+    include /etc/nginx/mime.types;
+    default_type application/octet-stream;
+
+    ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
+    ssl_prefer_server_ciphers on;
+
+    gzip on;
+    gzip_disable ""msie6"";
+
+    server {{
+        listen {originalUri.Port};
+        location / {{
+            proxy_pass {redirectUri};
+            proxy_http_version 1.1;
+            proxy_set_header Upgrade $http_upgrade;
+            proxy_set_header Host $host;
+            proxy_cache_bypass $http_upgrade;
+        }}
+    }}
+}}");
+
+            // stop Nginx so that we can restart it with a new config file
+            var startInfo = new ProcessStartInfo
+            {
+                FileName = "sudo",
+                Arguments = "nginx -s stop",
+                UseShellExecute = false,
+                CreateNoWindow = true,
+                RedirectStandardError = true,
+                RedirectStandardOutput = true,
+                // Trying a work around for https://github.com/aspnet/Hosting/issues/140.
+                RedirectStandardInput = true
+            };
+            
+            var stopNginx = new Process() { StartInfo = startInfo };
+            stopNginx.ErrorDataReceived += (sender, dataArgs) => { Logger.LogWarning(dataArgs.Data ?? string.Empty); };
```

Comment:
Can you put the process name in front of the message? It makes it easier to understand what's happening.